src/comments.js

```diff
@@ -871,6 +894,121 @@ function canAttachComment(node) {
+  const comment = commentPath.getValue();
+  const contents = printComment(commentPath, options);
+
+  if (!contents) {
+    return "";
+  }
+
+  const isBlock =
+    options.printer.isBlockComment && options.printer.isBlockComment(comment);
+
+  // Leading block comments should see if they need to stay on the
+  // same line or not.
+  if (isBlock) {
+    return concat([
+      contents,
+      hasNewline(options.originalText, options.locEnd(comment)) ? hardline : " "
+    ]);
+  }
+
+  return concat([contents, hardline]);
+}
+
+function printTrailingComment(commentPath, print, options) {
+  const comment = commentPath.getValue();
+  const contents = printComment(commentPath, options);
+  if (!contents) {
+    return "";
+  }
+  const isBlock =
+    options.printer.isBlockComment && options.printer.isBlockComment(comment);
+
+  if (
+    hasNewline(options.originalText, options.locStart(comment), {
+      backwards: true
+    })
+  ) {
+    // This allows comments at the end of nested structures:
+    // {
+    //   x: 1,
+    //   y: 2
+    //   // A comment
+    // }
+    // Those kinds of comments are almost always leading comments, but
+    // here it doesn't go "outside" the block and turns it into a
+    // trailing comment for `2`. We can simulate the above by checking
+    // if this a comment on its own line; normal trailing comments are
+    // always at the end of another expression.
+
+    const isLineBeforeEmpty = isPreviousLineEmpty(
+      options.originalText,
+      comment,
+      options
+    );
+
+    return lineSuffix(
+      concat([hardline, isLineBeforeEmpty ? hardline : "", contents])
+    );
+  } else if (isBlock) {
+    // Trailing block comments never need a newline
+    return concat([" ", contents]);
+  }
+
+  return concat([lineSuffix(` ${contents}`), !isBlock ? breakParent : ""]);
+}
+
+function printAllComments(path, print, options, needsSemi) {
+  const value = path.getValue();
+  const printed = print(path);
+  const comments = value && value.comments;
+
+  if (!comments || comments.length === 0) {
+    return prependCursorPlaceholder(path, options, printed);
+  }
+
+  const leadingParts = [];
+  const trailingParts = [needsSemi ? ";" : "", printed];
+
+  path.each(commentPath => {
+    const comment = commentPath.getValue();
+    const { leading, trailing } = comment;
+
+    if (leading) {
+      const contents = printLeadingComment(commentPath, print, options);
+      if (!contents) {
+        return;
+      }
+      leadingParts.push(contents);
+
+      const text = options.originalText;
+      if (hasNewline(text, skipNewline(text, options.locEnd(comment)))) {
+        leadingParts.push(hardline);
+      }
+    } else if (trailing) {
+      trailingParts.push(printTrailingComment(commentPath, print, options));
+    }
+  }, "comments");
+
+  return prependCursorPlaceholder(
+    path,
+    options,
+    concat(leadingParts.concat(trailingParts))
+  );
+}
```

Comment:
It is shame, but we need implement this api in prettier, right now i copy/paste code with little changes, i try to find on this issue in prettier in near future, now it is good solution